File: javascript/src/lib/math.js
/**
 * @param {number} n
 * @return {number}
 */
exports.factorial = function factorial(n) {
    let answer = 1;
    for (let x = 2; x <= n; x += 1) {
        answer *= x;
    }
    return answer;
};
/**
 * @param {number} n
 * @param {number} r
 * @return {number}
 */
exports.nChooseR = function nChooseR(n, r) {
    if (n <= 20) {
        return factorial(n) / factorial(r) / factorial(n - r);
    }

    let answer; let tmp; let i; let j;
    const factors = new Int8Array(n + 1);
    // collect factors of final number
    for (i = 2; i <= n; i += 1) {
        factors[i] = 1;
    }

    // negative factor values indicate need to divide
    for (i = 2; i <= r; i += 1) {
        factors[i] -= 1;
    }
    for (i = 2; i <= n - r; i += 1) {
        factors[i] -= 1;
    }

    // this loop reduces to prime factors only
    for (i = n; i > 1; i -= 1) {
        for (j = 2; j < i; j += 1) {
            if (i % j == 0) {
                factors[j] += factors[i];
                factors[i / j] += factors[i];
                factors[i] = 0;
                break;
            }
        }
    }

    i = j = 2;
    answer = 1;
    while (i <= n) {
        while (factors[i] > 0) {
            tmp = answer;
            answer *= i;
            while (answer < tmp && j <= n) {
                while (factors[j] < 0) {
                    tmp /= j;
                    factors[j] += 1;
                }
                j += 1;
                answer = tmp * i;
            }
            if (answer < tmp) {
                return ulong.MaxValue;
            } // this indicates an overflow
            factors[i] -= 1;
        }
        i += 1;
    }

    while (j <= n) {
        while (factors[j] < 0) {
            answer /= j;
            factors[j] += 1;
        }
        j += 1;
    }

    return answer;
};

File: lua/src/lib/math.lua
local function factorial(n)
    local answer = 1

    for i = 2,n do
        answer = answer * i
    end

    return answer
end

local function factor_gen(n, r)
    local factors = {}

    -- collect factors of final number
    for i = 2,n do
        factors[i] = 1
    end

    -- negative factor values indicate need to divide
    for i = 2,r do
        factors[i] = factors[i] - 1
    end
    for i = 2,(n - r) do
        factors[i] = factors[i] - 1
    end

    -- this loop reduces to prime factors only
    for i = n,2,-1 do
        for j = 2,(i - 1) do
            if i % j == 0 then
                factors[j] = factors[j] + factors[i]
                factors[i / j] = factors[i / j] + factors[i]
                factors[i] = 0
                break
            end
        end
    end

    return factors
end

local function n_choose_r(n, r)
    if n < 20 then -- fast path if number is small
        return factorial(n) / factorial(r) / factorial(n - r)
    end

    -- slow path for big numbers// slow path for larger numbers
    local factors = factor_gen(n, r)
    local answer
    local tmp
    local i = 2
    local j = 2
    answer = 1;
    while i <= n do
        while factors[i] > 0 do
            tmp = answer
            answer = answer * i
            while answer < tmp and j <= n do
                while factors[j] < 0 do
                    tmp = math.floor(tmp / j)
                    factors[j] = factors[j] + 1
                end
                j = j + 1;
                answer = tmp * i
            end

            if answer < tmp then
                return nil, "math error: overflow"
            end
            factors[i] = factors[i] - 1
        end
        i = i + 1
    end

    while j <= n do
        while factors[j] < 0 do
            answer = answer / j
            factors[j] = factors[j] + 1
        end
        j = j + 1
    end

    return answer
end

return {
    factorial = factorial,
    n_choose_r = n_choose_r,
}
